components/raftstore/src/store/fsm/apply.rs

```diff
@@ -1830,19 +1841,101 @@ where
             {
                 peer.set_id(*peer_id);
             }
-            write_peer_state(kv_wb_mut, &new_region, PeerState::Normal, None)
+            new_regions_map.insert(
+                new_region.get_id(),
+                (
+                    util::find_peer(&new_region, ctx.store_id).unwrap().get_id(),
+                    None,
+                ),
+            );
+            regions.push(new_region);
+        }
+
+        if right_derive {
+            derived.set_start_key(keys.pop_front().unwrap());
+            regions.push(derived.clone());
+        }
+
+        for (region_id, (_, reason)) in new_regions_map.iter_mut() {
+            let region_state_key = keys::region_state_key(*region_id);
+            match ctx
+                .engine
+                .get_msg_cf::<RegionLocalState>(CF_RAFT, &region_state_key)
+            {
+                Ok(None) => (),
+                Ok(Some(state)) => {
+                    *reason = Some(format!("state {:?} exist in kv engine", state));
+                }
+                e => panic!(
+                    "{} failed to get regions state of {}: {:?}",
+                    self.tag, region_id, e
+                ),
+            }
+        }
+
+        // Note that the following execution sequence is possible.
+        // Apply thread:    check `RegionLocalState`(None)
+        // Store thread:    create peer
+        // Peer thread:     apply snapshot and then be destroyed
+        // Apply thread:    check `StoreMeta` and find it's ok to create this new region.
+        // It's **very unlikely** to happen because the step 2 and step 3 should take far more time than the time interval
+        // between step 1 and step 4. (A similiar case can happen in create-peer process, see details in `maybe_create_peer`)
+        // Even it happens, this new region will be destroyed in future when it communicates to other TiKVs or PD.
+        // Now it seems there is no other side effects.
+        let mut meta = ctx.store_meta.lock().unwrap();
```

Comment:
Why not just lock it at L1859?

Reply:
Because it is very wasteful to hold the `StoreMeta` lock when reading `RegionLocalState` from disk.

Reply:
I think you can move L1885~L1912 to L1858, and cleanup metadata when conflict is detected. Then the corner case described in comment can't happen.